ostree-restore.sh

```diff
@@ -58,43 +39,44 @@ fi
 mount /sysroot -o remount,rw
 
 # Import OCIs
-log_it "Importing backup OCI"
-# Authentication is not working properly with unencapsulate
-# Will be fixed with https://github.com/ostreedev/ostree-rs-ext/pull/519
-# Until now, we will pull with podman and unencapsulate from local registry
-# ostree container unencapsulate --authfile "$my_dir/backup-secret.json" --repo /ostree/repo ostree-unverified-registry:$backup_refspec --write-ref $backup_tag
-podman_unencapsulate $backup_refspec $backup_tag "$my_dir/backup-secret.json"
-
-# If there's a parent to that commit, import it
-if [[ "$(ostree cat $backup_tag /rpm-ostree.json | jq -r '.deployments[] | select(.booted == true)| has("base-checksum")')" == "true" ]]; then
-    log_it "Parent commit found for base, importing OCI"
-    # ostree container unencapsulate --authfile "$my_dir/backup-secret.json" --repo /ostree/repo ostree-unverified-registry:$parent_refspec --write-ref $parent_tag
-    podman_unencapsulate $parent_refspec $parent_tag "$my_dir/backup-secret.json"
-fi
+log_it "Pulling and mounting backup OCI"
+podman pull --authfile="$my_dir/backup-secret.json" ${oneimage_refspec}
+img_mnt=$(podman image mount ${oneimage_refspec})
 
-log_it "Initializing and deploying new stateroot"
-if rpm-ostree status -v --json | jq -e '[.deployments[] | select(.osname == "ibu")] | length == 0' > /dev/null; then
-    ostree admin os-init $new_osname
-else
-    echo "Deployment already exists, skipping os-init"
-fi
-ostree container image deploy --sysroot / --stateroot $new_osname $(build_kargs) --authfile "$my_dir/backup-secret.json" --imgref ostree-unverified-registry:$base_refspec
-ostree_deploy=$(ostree admin status | awk /$new_osname/'{print $2}')
+ostree_repo=$(mktemp -d -p /var/tmp)
 
-# Workaround to fix deploy origin URL
-log_it "Restoring original osImageURL to new stateroot origin"
-original_osimage=$(ostree cat backup /mco-currentconfig.json | jq -r .spec.osImageURL)
-sed -e "s%docker://.*%$original_osimage%g" -i /ostree/deploy/$new_osname/deploy/$ostree_deploy.origin
+tar xzf $img_mnt/ostree.tgz --selinux -C $ostree_repo
+
+# Collect seed deployment data from the backup
+upg_booted_id=$(jq -r '.deployments[] | select(.booted == true) | .id' ${img_mnt}/rpm-ostree.json)
+upg_booted_deployment=${upg_booted_id/*-}
+upg_booted_ref=${upg_booted_deployment/\.*}
+
+up_ver=$(jq -r '.status.desired.version' ${img_mnt}/clusterversion.json)
+new_osname=rhcos_${up_ver}
+
+log_it "Importing remote ostree"
+ostree pull-local $ostree_repo
+ostree admin os-init $new_osname
+
+log_it "Creating new deployment $new_osname"
+# We should create the new deploy as not-default, and after the whole process is done, be able to switch to it for the next reboot
+# ostree admin deploy --os ${new_osname} $(build_kargs) --not-as-default ${upg_booted_ref}
+# Until I find how to do the switch, I'll deploy as default
```

Comment:
Yes, will need to see about getting https://github.com/ostreedev/ostree/blob/main/man/ostree-admin-set-default.xml shipped back to ocp 4.14 I guess.

Reply:
Yeah, we will need that functionality in 4.14 for sure

We'd like to have it in a way that if something happens that might make the node reboot between the new os is deployed and when we're ready for upgrade, the system will still boot properly with the "old version"